In [61]:
import pandas as pd

In [62]:
df = pd.read_csv('2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv')

## Data exploration

In [63]:
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [64]:
df.isnull().sum()

order_id          0
shop_id           0
user_id           0
order_amount      0
total_items       0
payment_method    0
created_at        0
dtype: int64

In [65]:
df.shape

(5000, 7)

In [66]:
df[['order_amount','total_items']].describe()

order_amount  total_items
count    5000.000000   5000.00000
mean     3145.128000      8.78720
std     41282.539349    116.32032
min        90.000000      1.00000
25%       163.000000      1.00000
50%       284.000000      2.00000
75%       390.000000      3.00000
max    704000.000000   2000.00000

In [67]:
df['shop_id'].value_counts()

53    68
71    66
19    64
13    63
89    61
      ..
74    38
56    37
67    37
38    35
85    35
Name: shop_id, Length: 100, dtype: int64

In [69]:
by_order_amount = df.groupby('order_amount').order_id.count().reset_index()
by_order_amount

order_amount  order_id
0              90        18
1              94        25
2             101        15
3             111        16
4             112        48
..            ...       ...
253         51450        16
254         77175         9
255        102900         1
256        154350         1
257        704000        17

[258 rows x 2 columns]

In [75]:
## Creating a new data frame to check the total sales and item amount per store

In [45]:
def compute_agg_stats(df):
    column_names = ['shop_id', 'total_order', 'total_items']
    by_store = df.groupby('shop_id').agg(total_order = ('order_amount', 'sum'), total_items = ('total_items', 'sum')).reset_index()
    new_df = pd.DataFrame(by_store, columns = column_names).sort_values('total_order', ascending=False)
    return new_df

In [46]:
new_df = compute_agg_stats(df)

In [47]:
## Creating a new column to show one pair of snickers price for the store

In [48]:
new_df['pair_price'] = new_df['total_order'] / new_df['total_items']

In [51]:
new_df.sort_values('pair_price', ascending=False)

shop_id  total_order  total_items  pair_price
77       78      2263800           88     25725.0
41       42     11990176        34063       352.0
11       12        18693           93       201.0
88       89        23128          118       196.0
98       99        18330           94       195.0
..      ...          ...          ...         ...
6         7        12208          109       112.0
99      100         8547           77       111.0
31       32         7979           79       101.0
1         2         9588          102        94.0
91       92         6840           76        90.0

[100 rows x 4 columns]

In [53]:
new_df.describe()

shop_id   total_order   total_items    pair_price
count  100.000000  1.000000e+02    100.000000    100.000000
mean    50.500000  1.572564e+05    439.360000    407.990000
std     29.011492  1.216218e+06   3396.366111   2557.462906
min      1.000000  6.840000e+03     67.000000     90.000000
25%     25.750000  1.293050e+04     88.000000    132.750000
50%     50.500000  1.488750e+04    100.000000    153.000000
75%     75.250000  1.760000e+04    111.250000    168.250000
max    100.000000  1.199018e+07  34063.000000  25725.000000

In [59]:
## Getting rid of outliers in original data
df = df[df['shop_id'] != 78]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[4954 rows x 7 columns]

## Looking for a better metric

In [60]:
df['order_amount'].median()

284.0

## Answers:

1a. 
Using only mean to compute the AOV can be very dangerous, because mean metric is very sensitive to extreme values (outliers).
In our case you can see, that one of the stores (shop_id = 78) was reporting 25725 sale price per item, that made it's orders amounts very high.
Also we can see that we have report of 704000 amount per transaction 17 time, these transactions are also rising up the AOV. 

1b. 
I think the median would be a better choise in this case, because it is less affected by extreme values.

1c.
Median that I got is 284, that looks more reasonable than the mean value that we got before.

In [72]:
'''
2a.

Answer = 54

SELECT COUNT(OrderID) FROM Orders
WHERE ShipperID = (
SELECT ShipperID FROM Shippers
WHERE ShipperName = 'Speedy Express')
'''

"\n2a.\n\nAnswer = 54\n\nSELECT COUNT(OrderID) FROM Orders\nWHERE ShipperID = (\nSELECT ShipperID FROM Shippers\nWHERE ShipperName = 'Speedy Express')\n"

In [73]:
'''
2b.

Answer = Peacock

SELECT LastName FROM Employees
WHERE EmployeeID = (
SELECT EmployeeID
FROM Orders
GROUP BY EmployeeID
ORDER BY COUNT(OrderID) DESC
LIMIT 1
);
'''

'\n2b.\n\nAnswer = Peacock\n\nSELECT LastName FROM Employees\nWHERE EmployeeID = (\nSELECT EmployeeID\nFROM Orders\nGROUP BY EmployeeID\nORDER BY COUNT(OrderID) DESC\nLIMIT 1\n);\n'

In [74]:
'''
2c.

Answer = Boston Crab Meat

SELECT p.ProductName, SUM(Quantity) as TotalQuantity
FROM Orders as o
JOIN OrderDetails as od ON od.OrderID = o.OrderID
JOIN Customers as c ON c.CustomerID = o.CustomerID
JOIN Products as p ON od.ProductID = p.ProductID
WHERE c.Country = "Germany"
GROUP BY p.ProductID
ORDER BY TotalQuantity DESC
LIMIT 1;
'''

'\n2c.\n\nAnswer = Boston Crab Meat\n\nSELECT p.ProductName, SUM(Quantity) as TotalQuantity\nFROM Orders as o\nJOIN OrderDetails as od ON od.OrderID = o.OrderID\nJOIN Customers as c ON c.CustomerID = o.CustomerID\nJOIN Products as p ON od.ProductID = p.ProductID\nWHERE c.Country = "Germany"\nGROUP BY p.ProductID\nORDER BY TotalQuantity DESC\nLIMIT 1;\n'